nodejs/lancedb/embedding/embedding_function.ts

```diff
@@ -12,67 +16,138 @@
-   * to an empty table, then that process will fail.
+   * @example
+   * ```ts
+   * class MyEmbeddingFunction extends EmbeddingFunction {
+   *   constructor(options: {model: string, timeout: number}) {
+   *     super();
+   *     this.model = options.model;
+   *     this.timeout = options.timeout;
+   *   }
+   *   toJSON() {
+   *     return {
+   *       model: this.model,
+   *       timeout: this.timeout,
+   *     };
+   * }
+   * ```
    */
-  embeddingDimension?: number;
+  abstract toJSON(): Partial<M>;
 
   /**
-   * The name of the column that will contain the embedding
+   * sourceField is used in combination with `LanceSchema` to provide a declarative data model
    *
-   * By default this is "vector"
+   * @param optionsOrDatatype - The options for the field or the datatype
+   *
+   * @see {@link lancedb.LanceSchema}
    */
-  destColumn?: string;
+  sourceField(
+    options: Partial<FieldOptions> | DataType,
+  ): [DataType, Map<string, EmbeddingFunction>] {
+    const datatype = options instanceof DataType ? options : options?.datatype;
+    if (!datatype) {
+      throw new Error("Datatype is required");
+    }
+    const metadata = new Map<string, EmbeddingFunction>();
+    metadata.set("source_column_for", this);
+
+    if (options instanceof DataType) {
+      return [options, metadata];
+    } else {
+      return [datatype, metadata];
+    }
+  }
 
   /**
-   * Should the source column be excluded from the resulting table
+   * vectorField is used in combination with `LanceSchema` to provide a declarative data model
    *
-   * By default the source column is included.  Set this to true and
-   * only the embedding will be stored.
+   * @param options - The options for the field
+   *
+   * @see {@link lancedb.LanceSchema}
    */
-  excludeSource?: boolean;
+  vectorField(
+    options?: Partial<FieldOptions>,
+  ): [DataType, Map<string, EmbeddingFunction>] {
+    let dtype: DataType;
+    const dims = this.ndims() ?? options?.dims;
+    if (!options?.datatype) {
+      if (dims === undefined) {
+        throw new Error("ndims is required for vector field");
+      }
+      dtype = new FixedSizeList(dims, new Field("item", new Float32()));
+    } else {
+      if (options.datatype instanceof FixedSizeList) {
+        dtype = options.datatype;
+      } else if (options.datatype instanceof Float) {
+        if (dims === undefined) {
+          throw new Error("ndims is required for vector field");
+        }
+        dtype = newVectorType(dims, options.datatype);
+      } else {
+        throw new Error(
+          "Expected FixedSizeList or Float as datatype for vector field",
+        );
+      }
+    }
+    const metadata = new Map<string, EmbeddingFunction>();
+    metadata.set("vector_column_for", this);
+
+    return [dtype, metadata];
+  }
+
+  /** The number of dimensions of the embeddings */
+  ndims(): number | undefined {
+    return undefined;
+  }
```

Comment:
Should this just be abstract?

Reply:
if it's abstract, it requires the user to provide an implementation, even if the function can return `undefined`. This just provides a default implementation so users don't need to implement it if they dont want.